### PR TITLE
Add file containing release notes, that currently only include the re…

### PR DESCRIPTION
**Pull request by keyboardDrummer**
Add file containing release notes, that currently only include the release notes for 3.4

This files makes it easier to review release notes for upcoming releases. Possibly we should put release notes from past releases into it as well

### CODE REVIEW

File: RELEASE_NOTES.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# 3.4 
+
+The top four improvements in Dafny 3.4 are:
+- For certain classes of changes, prevent unexpected changes in verification behavior when changing a Dafny program 
+- Add command line options to assist in debugging verification performance.
+- Critical fixes to the IDE and greatly improved responsiveness of non-verification IDE features. 
+- The C# back-end supports traits as type parameters on datatypes.
```
Comment (RustanLeino) [inline RELEASE_NOTES.md:7]: Add a note that compiled downcasts are supported for these only for co-variant and non-variant type parameters, not yet contra-variant type parameters. This same note also applies under the `C#` section below (or perhaps this detail should only be mentioned in the `C#` section below, keeping this list of "top improvements" simpler to read).

**Review by RustanLeino — CHANGES_REQUESTED** [review]

File: RELEASE_NOTES.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+- feat: Improve the hover feature to work for a wider scope of Dafny constructs, including function and method parameters, forall, exists and let expressions, and set and map comprehensions.
+- feat: Add an experimental verification caching feature, which enables automatically determining which proofs need to verify again after making changes.
+
+### UX
+- feat: Improve error reporting when providing wrongly typed arguments in a function call
+- feat: Improve error reporting when using type tests
+
+### C#
+- fix: resolve an instance where Dafny would produce invalid C# code: https://github.com/dafny-lang/dafny/issues/1607
+- feat: support traits as type parameters on datatypes for co-variant and non-variant type parameter, but not for contra-variant type parameters: https://github.com/dafny-lang/dafny/issues/1499
```
Comment (fabiomadge) [inline RELEASE_NOTES.md:34]:
```suggestion
- feat: support for variant type parameters on datatype definitions, with the omission of downcasting ones with conta-variant type parameters, enabling
traits as type arguments: https://github.com/dafny-lang/dafny/pull/1645
```
Please include the changes in #1645.

**Review by fabiomadge — CHANGES_REQUESTED** [review]

File: RELEASE_NOTES.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+## Breaking changes
+
+- Proofs such as methods and lemmas whose verification behavior, either the result or the verification time, depend on arbitrary behavior in Dafny's solver, may show different verification behavior in Dafny 3.4. For customers who are not ready to change those proofs to make their verification more reliable, we offer the command line option `/mimicVerificationOf:3.3` to keep the Dafny 3.4 verification behavior consistent with 3.3.
```
Comment (fabiomadge) [inline RELEASE_NOTES.md:38]: Is _customers_ the language we want here?

File: RELEASE_NOTES.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
+# 3.4 
+
+The top four improvements in Dafny 3.4 are:
```
Comment (fabiomadge) [inline RELEASE_NOTES.md:3]: I don't think we need this line.

File: RELEASE_NOTES.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+- fix: resolve an instance where Dafny would produce invalid C# code: https://github.com/dafny-lang/dafny/issues/1607
+- feat: support traits as type parameters on datatypes for co-variant and non-variant type parameter, but not for contra-variant type parameters: https://github.com/dafny-lang/dafny/issues/1499
+
+## Breaking changes
+
+- Proofs such as methods and lemmas whose verification behavior, either the result or the verification time, depend on arbitrary behavior in Dafny's solver, may show different verification behavior in Dafny 3.4. For customers who are not ready to change those proofs to make their verification more reliable, we offer the command line option `/mimicVerificationOf:3.3` to keep the Dafny 3.4 verification behavior consistent with 3.3.
+
+- In Dafny 3.3, comprehensions quantified over subset types did not validate the constraint of the subset type, which could result in crashes at run-time. In 3.4, subset types are disabled in set comprehensions in compiled contexts, unless the subset constraint is itself compilable.
+
+Before, the following code would pass Dafny and be compiled without error, but would crash at run-time:
```
Comment (cpitclaudel) [inline RELEASE_NOTES.md:42]: This part should be indented if you want it to be part of the bullet point above

File: RELEASE_NOTES.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  var t = set x: RefinedData | x in s;
+  forall x in t {
+    if !ghostFunction(x) {
+      var crash := 1/0;
+    }
+  }
+}
+```
+In Dafny 3.4, the same code triggers a resolution error of the form:
+Error: RefinedData is a subset type and its constraint is not compilable, hence it cannot yet be used as the type of a bound variable in set comprehension. The next error will explain why the constraint is not compilable.
```
Comment (cpitclaudel) [inline RELEASE_NOTES.md:56]: This could be a quote or a code block

File: RELEASE_NOTES.md (hunk 4 above)
Comment (keyboardDrummer) [reply RELEASE_NOTES.md:3]: I'm inclined to agree. I modelled this on the 3.3 release notes that do include a summary section, although they're significantly bigger in total: https://github.com/dafny-lang/dafny/releases/tag/v3.3.0

Still think we should leave it out?

File: RELEASE_NOTES.md (hunk 2 above)
Comment (keyboardDrummer) [reply RELEASE_NOTES.md]: I've left out the PR reference, since the rest of the notes don't reference PRs either. I think it makes more sense to reference issue than PRs, don't you? The issues can then reference the PRs again.

File: RELEASE_NOTES.md (hunk 4 above)
Comment (fabiomadge) [reply RELEASE_NOTES.md:3]: I like the headline features, but I don't think they need an introduction.

File: RELEASE_NOTES.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
+# 3.4 
+
+The top four improvements in Dafny 3.4 are:
+- For certain classes of changes, prevent unexpected changes in verification behavior when changing a Dafny program 
+- Add command line options to assist in debugging verification performance.
+- Critical fixes to the IDE and greatly improved responsiveness of non-verification IDE features. 
+- The C# back-end partially supports traits as type parameters on datatypes.
```
Comment (fabiomadge) [inline RELEASE_NOTES.md:7]:
```suggestion
- The C# back-end supports traits as type parameters on datatypes.
```

File: RELEASE_NOTES.md [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+- feat: Improve the hover feature to work for a wider scope of Dafny constructs, including function and method parameters, forall, exists and let expressions, and set and map comprehensions.
+- feat: Add an experimental verification caching feature, which enables automatically determining which proofs need to verify again after making changes.
+
+### UX
+- feat: Improve error reporting when providing wrongly typed arguments in a function call
+- feat: Improve error reporting when using type tests
+
+### C#
+- fix: resolve an instance where Dafny would produce invalid C# code. https://github.com/dafny-lang/dafny/issues/1607
+- feat: support variant type parameters on datatype definitions, which enables using traits as type arguments. Support for downcasting datatypes with contra-variant type parameters is left out. https://github.com/dafny-lang/dafny/issues/1499
```
Comment (fabiomadge) [inline RELEASE_NOTES.md:34]:
```suggestion
- feat: Support variant type parameters on datatype definitions, which enables using traits as type arguments. https://github.com/dafny-lang/dafny/issues/1499
- feat: Support for downcasting both custom datatypes and functions.
```

**Review by fabiomadge — CHANGES_REQUESTED** [review]

**Review by fabiomadge — DISMISSED** [review]

File: RELEASE_NOTES.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
+# 3.4 
+
+The top four improvements in Dafny 3.4 are:
+- For certain classes of changes, prevent unexpected changes in verification behavior when changing a Dafny program.
```
Comment (robin-aws) [inline RELEASE_NOTES.md:4]:
```suggestion
- For certain classes of changes to a Dafny program, prevent unexpected changes in verification behavior.
```

Just to avoid the slight redundancy.

File: RELEASE_NOTES.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+- Add command line options to assist in debugging verification performance.
+- Critical fixes to the IDE and greatly improved responsiveness of non-verification IDE features. 
+- The C# back-end supports traits as type parameters on datatypes.
+
+### Verification
+- fix: Resolve unsoundness issue (https://github.com/dafny-lang/dafny/issues/1619).
+- fix: Don't silently succeed if the solver crashes (https://github.com/boogie-org/boogie/pull/488).
+- feat: Prevent changes in the verification behavior of a proof, when any of these types of changes are made to Dafny user code:
+  - Changes logically unrelated to the proof being verified
```
Comment (robin-aws) [inline RELEASE_NOTES.md:13]:
```suggestion
  - Changes to declarations not referenced by the method being verified
```

The original claim is too strong, since it's not yet true for unrelated changes within the same method - @keyboardDrummer or @RustanLeino please confirm if I've over-simplified in my version.

File: RELEASE_NOTES.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+  - Changes to the name of any declaration
+  - Changes to the order of top-level declarations
+- feat: Assist in debugging the verification performance of a proof by adding the `/vcsSplitOnEveryAssert` CLI option and `{:vcs_split_on_every_assert}` attribute, and report the outcome and duration of splits when they occur in `/verificationLogger:trx` content.
```
Comment (robin-aws) [inline RELEASE_NOTES.md:16]:
```suggestion
- feat: Assist in debugging the verification performance of a proof by adding the `/vcsSplitOnEveryAssert` CLI option and `{:vcs_split_on_every_assert}` attribute (see https://github.com/boogie-org/boogie/issues/465), and report the outcome and duration of splits when they occur in `/verificationLogger:trx` content.
```

Seems fair to note that the first part is actually a Boogie feature (especially since we'll want to make it a much more polished Dafny feature soon: https://github.com/dafny-lang/dafny/issues/1754)

File: RELEASE_NOTES.md [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+- feat: Add a `/verificationLogger:csv` CLI option that emits the same status and timing information as `/verificationLogger:trx`, but in an easier-to-parse format, along with Z3 resource counts for more repeatable tracking of verification difficulty.
+
+### IDE 
+- fix: Clean up process resources if IDE closed or restarted.
+- fix: Do not let the Dafny compilation status bar get in a stuck state.
+
```
Comment (robin-aws) [inline RELEASE_NOTES.md:22]: Minor: I slightly prefer listing the features before the bug fixes - it's a little more optimistic :)

File: RELEASE_NOTES.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+- feat: Verification status reporting shows which proof is being verified, which can help debug slow to verify proofs.
+- feat: Publish parsing and resolution diagnostics before verification has completed. Verification diagnostics from previous runs are migrated.
+- feat: Enable 'go to definition', 'hover' and 'signature help' features before verification has completed.
+- feat: Improve the hover feature to work for a wider scope of Dafny constructs, including function and method parameters, forall, exists and let expressions, and set and map comprehensions.
+- feat: Add an experimental verification caching feature, which enables automatically determining which proofs need to verify again after making changes.
+- feat: Display related resolution errors using nested diagnostics instead of independent diagnostics.
+
+### UX
+- feat: Improve error reporting when providing incorrectly typed arguments in a function call.
```
Comment (robin-aws) [inline RELEASE_NOTES.md:31]:
```suggestion
- feat: Improve error reporting when providing incorrectly-typed arguments in a function call.
```

File: RELEASE_NOTES.md [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+- feat: Improve error reporting when using type tests.
+
+### C#
+- fix: resolve various instances where Dafny would produce invalid C# code (#1607, #1761, and #1762).
```
Comment (robin-aws) [inline RELEASE_NOTES.md:35]:
```suggestion
- fix: Resolve various instances where Dafny would produce invalid C# code (https://github.com/dafny-lang/dafny/issues/1607, https://github.com/dafny-lang/dafny/issues/1761, and https://github.com/dafny-lang/dafny/issues/1762).
```

File: RELEASE_NOTES.md [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+- feat: Support variant type parameters on datatype definitions, which enables using traits as type arguments (https://github.com/dafny-lang/dafny/issues/1499).
+- feat: Support for downcasting both custom datatypes and functions.
```
Comment (robin-aws) [inline RELEASE_NOTES.md:37]: Could we be more specific and/or link the issue here? The description is a little ambiguous - does it mean explicit `is` tests, or implicit downcasting because the verifier proved no `is` is necessary?

File: RELEASE_NOTES.md [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+### Various improvements
+- fix: `DafnyLanguageServer.dll` and `Dafny.dll` depended on two different versions of Newtonsoft.Json, which could cause crashes in development environments.
+- fix: (error reporting) Types with the same name but from different modules are now disambiguated in error messages.
+- fix: (error reporting) Messages about arguments / parameters type mismatch are clearer and include the parameter name if available.
+- fix: (robustness) Exceptions during parsing, if any, won't crash the language server anymore.
+- fix: The elephant operator (`:-`) has a clearer error message and no longer reject generic methods on its right-hand side.
+
+## Breaking changes
+
+- Proofs such as methods and lemmas whose verification behavior, either the result or the verification time, depend on arbitrary behavior in Dafny's solver, may show different verification behavior in Dafny 3.4. For users who are not ready to change those proofs to make their verification more reliable, we offer the command line option `/mimicVerificationOf:3.3` to keep the Dafny 3.4 verification behavior consistent with 3.3.
```
Comment (robin-aws) [inline RELEASE_NOTES.md:48]:
```suggestion
- The verifier in Dafny 3.4 is now more efficient for many programs, and making changes to Dafny programs is less likely to cause verification to take longer or timeout. However, it is still possible for some correct programs to take longer to verify than on Dafny 3.3, or for verification to fail. For users with such programs who are not yet ready to modify them to pass the 3.4 verifier, we offer the command line option `/mimicVerificationOf:3.3` to keep the Dafny 3.4 verification behavior consistent with 3.3.
```

Saying the verification result can depend on arbitrary behavior in Dafny's solver seems to imply that invalid code might successfully verify.

File: RELEASE_NOTES.md (hunk 4 above)
Comment (robin-aws) [reply RELEASE_NOTES.md:3]: I definitely like highlighting the features at least.  I don't have a strong opinion about this exact line.

**Review by robin-aws — CHANGES_REQUESTED** [review]

File: RELEASE_NOTES.md (hunk 15 above)
Comment (atomb) [reply RELEASE_NOTES.md:37]: Probably for @fabiomadge to answer?

File: RELEASE_NOTES.md (hunk 10 above)
Comment (RustanLeino) [reply RELEASE_NOTES.md:13]: Yeah, I like the updated wording, too.

File: RELEASE_NOTES.md [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+
+- feat: Verification status reporting shows which proof is being verified, which can help debug slow to verify proofs.
+- feat: Publish parsing and resolution diagnostics before verification has completed. Verification diagnostics from previous runs are migrated.
+- feat: Enable 'go to definition', 'hover' and 'signature help' features before verification has completed.
+- feat: Improve the hover feature to work for a wider scope of Dafny constructs, including function and method parameters, forall, exists and let expressions, and set and map comprehensions.
+- feat: Add an experimental verification caching feature, which enables automatically determining which proofs need to verify again after making changes.
+- feat: Display related resolution errors using nested diagnostics instead of independent diagnostics.
+
+### UX
+- feat: Improve error reporting when providing incorrectly-typed arguments in a function call.
```
Comment (RustanLeino) [inline RELEASE_NOTES.md:31]:
```suggestion
- feat: Improve error reporting when providing incorrectly typed arguments in a function call.
```

File: RELEASE_NOTES.md (hunk 15 above)
Comment (fabiomadge) [reply RELEASE_NOTES.md:37]: This should affect both explicit `as` casts and implicit ones. A runtime check using `is` is unrelated.

**Review by robin-aws — APPROVED** [review]
